File: port-scanner.py
import socket
import os
black = "\033[30m" 
red = "\033[31m"
green = "\033[32m" 
yellow = "\033[33m" 
blue = "\033[34m" 
magenta = "\033[35m" 
cyan = "\033[36m" 
white = "\033[37m" 
orange = "\033[38;5;208m"
cosmic = "\033[38;2;255;102;102m" 
reset = "\033[0m"
os.system("clear")
print(orange + '''

 ██░ ██  ▄▄▄       ▄████▄   ██ ▄█▀▓█████  ██▀███  ▄▄▄█████▓ ▒█████   ▒█████   ██▓    
▓██░ ██▒▒████▄    ▒██▀ ▀█   ██▄█▒ ▓█   ▀ ▓██ ▒ ██▒▓  ██▒ ▓▒▒██▒  ██▒▒██▒  ██▒▓██▒    
▒██▀▀██░▒██  ▀█▄  ▒▓█    ▄ ▓███▄░ ▒███   ▓██ ░▄█ ▒▒ ▓██░ ▒░▒██░  ██▒▒██░  ██▒▒██░    
░▓█ ░██ ░██▄▄▄▄██ ▒▓▓▄ ▄██▒▓██ █▄ ▒▓█  ▄ ▒██▀▀█▄  ░ ▓██▓ ░ ▒██   ██░▒██   ██░▒██░    
░▓█▒░██▓ ▓█   ▓██▒▒ ▓███▀ ░▒██▒ █▄░▒████▒░██▓ ▒██▒  ▒██▒ ░ ░ ████▓▒░░ ████▓▒░░██████▒
 ▒ ░░▒░▒ ▒▒   ▓▒█░░ ░▒ ▒  ░▒ ▒▒ ▓▒░░ ▒░ ░░ ▒▓ ░▒▓░  ▒ ░░   ░ ▒░▒░▒░ ░ ▒░▒░▒░ ░ ▒░▓  ░
 ▒ ░▒░ ░  ▒   ▒▒ ░  ░  ▒   ░ ░▒ ▒░ ░ ░  ░  ░▒ ░ ▒░    ░      ░ ▒ ▒░   ░ ▒ ▒░ ░ ░ ▒  ░
 ░  ░░ ░  ░   ▒   ░        ░ ░░ ░    ░     ░░   ░   ░      ░ ░ ░ ▒  ░ ░ ░ ▒    ░ ░   
 ░  ░  ░      ░  ░░ ░      ░  ░      ░  ░   ░                  ░ ░      ░ ░      ░  ░
                  ░                                                                  
  ██████  ▄████▄   ▄▄▄       ███▄    █  ███▄    █ ▓█████  ██▀███                     
▒██    ▒ ▒██▀ ▀█  ▒████▄     ██ ▀█   █  ██ ▀█   █ ▓█   ▀ ▓██ ▒ ██▒                   
░ ▓██▄   ▒▓█    ▄ ▒██  ▀█▄  ▓██  ▀█ ██▒▓██  ▀█ ██▒▒███   ▓██ ░▄█ ▒                   
  ▒   ██▒▒▓▓▄ ▄██▒░██▄▄▄▄██ ▓██▒  ▐▌██▒▓██▒  ▐▌██▒▒▓█  ▄ ▒██▀▀█▄                     
▒██████▒▒▒ ▓███▀ ░ ▓█   ▓██▒▒██░   ▓██░▒██░   ▓██░░▒████▒░██▓ ▒██▒                   
▒ ▒▓▒ ▒ ░░ ░▒ ▒  ░ ▒▒   ▓▒█░░ ▒░   ▒ ▒ ░ ▒░   ▒ ▒ ░░ ▒░ ░░ ▒▓ ░▒▓░                   
░ ░▒  ░ ░  ░  ▒     ▒   ▒▒ ░░ ░░   ░ ▒░░ ░░   ░ ▒░ ░ ░  ░  ░▒ ░ ▒░                   
░  ░  ░  ░          ░   ▒      ░   ░ ░    ░   ░ ░    ░     ░░   ░                    
      ░  ░ ░            ░  ░         ░          ░    ░  ░   ░                        
         ░                                                                           
''' + orange)
print(cosmic + "tool by hackertool" + cosmic)
print()
def scan_port(ip, port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(0.2) # if there is any problem or the code can not check if the port open change it to 0.5 or 0.4 or 0.3 or 1
    result = sock.connect_ex((ip, port)) # if that's have failed to connection with the ip and port 
    if result == 0: # if the reslut 0 will be opend port 
        print(f"{green}Port {port} is opened{green}")
    else:
        print(f"{red}Port {port} is closed{red}")
    sock.close()

def scan_ports(ip, start_port, end_port):
    for port in range(start_port, end_port + 1):
        scan_port(ip, port)

if __name__ == "__main__":
    target = input(f"{green}type the ip or webstie : {green}")
    start_port = int(input("type the starting port : "))
    end_port = int(input("type the ending port : "))
    print(f"{orange} scanning {target} from port {start_port} to {end_port}...{orange}")
    print(f"{magenta} note if the code can't check if the target is open change  it to 0.5 or 0.3 or 0.4 or 1 {magenta}")
    scan_ports(target, start_port, end_port)

File: port-scanner.rb
require 'socket' #require socket
require 'timeout' #require timeout
#type colors ANSCI escape code
$black = "\033[30m" 
$red = "\033[31m"
$green = "\033[32m" 
$yellow = "\033[33m" 
$blue = "\033[34m" 
$magenta = "\033[35m" 
$cyan = "\033[36m" 
$white = "\033[37m" 
$orange = "\033[38;5;208m" 
$arctic = "\033[38;2;135;206;250m"
$reset = "\033[0m"
system("clear") #clear
#orange banner
puts $orange + '''

 ██░ ██  ▄▄▄       ▄████▄   ██ ▄█▀▓█████  ██▀███  ▄▄▄█████▓ ▒█████   ▒█████   ██▓    
▓██░ ██▒▒████▄    ▒██▀ ▀█   ██▄█▒ ▓█   ▀ ▓██ ▒ ██▒▓  ██▒ ▓▒▒██▒  ██▒▒██▒  ██▒▓██▒    
▒██▀▀██░▒██  ▀█▄  ▒▓█    ▄ ▓███▄░ ▒███   ▓██ ░▄█ ▒▒ ▓██░ ▒░▒██░  ██▒▒██░  ██▒▒██░    
░▓█ ░██ ░██▄▄▄▄██ ▒▓▓▄ ▄██▒▓██ █▄ ▒▓█  ▄ ▒██▀▀█▄  ░ ▓██▓ ░ ▒██   ██░▒██   ██░▒██░    
░▓█▒░██▓ ▓█   ▓██▒▒ ▓███▀ ░▒██▒ █▄░▒████▒░██▓ ▒██▒  ▒██▒ ░ ░ ████▓▒░░ ████▓▒░░██████▒
 ▒ ░░▒░▒ ▒▒   ▓▒█░░ ░▒ ▒  ░▒ ▒▒ ▓▒░░ ▒░ ░░ ▒▓ ░▒▓░  ▒ ░░   ░ ▒░▒░▒░ ░ ▒░▒░▒░ ░ ▒░▓  ░
 ▒ ░▒░ ░  ▒   ▒▒ ░  ░  ▒   ░ ░▒ ▒░ ░ ░  ░  ░▒ ░ ▒░    ░      ░ ▒ ▒░   ░ ▒ ▒░ ░ ░ ▒  ░
 ░  ░░ ░  ░   ▒   ░        ░ ░░ ░    ░     ░░   ░   ░      ░ ░ ░ ▒  ░ ░ ░ ▒    ░ ░   
 ░  ░  ░      ░  ░░ ░      ░  ░      ░  ░   ░                  ░ ░      ░ ░      ░  ░
                  ░                                                                  
  ██████  ▄████▄   ▄▄▄       ███▄    █  ███▄    █ ▓█████  ██▀███                     
▒██    ▒ ▒██▀ ▀█  ▒████▄     ██ ▀█   █  ██ ▀█   █ ▓█   ▀ ▓██ ▒ ██▒                   
░ ▓██▄   ▒▓█    ▄ ▒██  ▀█▄  ▓██  ▀█ ██▒▓██  ▀█ ██▒▒███   ▓██ ░▄█ ▒                   
  ▒   ██▒▒▓▓▄ ▄██▒░██▄▄▄▄██ ▓██▒  ▐▌██▒▓██▒  ▐▌██▒▒▓█  ▄ ▒██▀▀█▄                     
▒██████▒▒▒ ▓███▀ ░ ▓█   ▓██▒▒██░   ▓██░▒██░   ▓██░░▒████▒░██▓ ▒██▒                   
▒ ▒▓▒ ▒ ░░ ░▒ ▒  ░ ▒▒   ▓▒█░░ ▒░   ▒ ▒ ░ ▒░   ▒ ▒ ░░ ▒░ ░░ ▒▓ ░▒▓░                   
░ ░▒  ░ ░  ░  ▒     ▒   ▒▒ ░░ ░░   ░ ▒░░ ░░   ░ ▒░ ░ ░  ░  ░▒ ░ ▒░                   
░  ░  ░  ░          ░   ▒      ░   ░ ░    ░   ░ ░    ░     ░░   ░                    
      ░  ░ ░            ░  ░         ░          ░    ░  ░   ░                        
         ░                                                                           
''' + $orange
puts $arctic + "tool by hackertool " + $arctic
puts
def scan_port(ip, port) #def using
  begin
    Timeout.timeout(0.4) do #timeout of scan
      socket = TCPSocket.new(ip, port) # connection to ip and port to check if was closed or open
      puts $green + "port #{port} is opened" + $green #print port is open with port name in green color
      socket.close # close 
    end 
  rescue Timeout::Error, Errno::ECONNREFUSED # that's have failed to connection with the ip and port 
    puts $red + "port #{port} is closed" + $red # print port is closed with port name in red color
  rescue Errno::EHOSTUNREACH 
    # If the host is unreachable
    puts $orange + "the host #{ip} you type it is unreachable" + $orange
  rescue => e
    # Catch other errors
    puts $red + "error scanning port #{port}: #{e.message}" + $red
  end
end

def scan_ports(ip, start_port, end_port)
  (start_port..end_port).each do |port|
    scan_port(ip, port)
  end
end

print $blue + "type the ip or website name : " + $blue
target_ip = gets.chomp
print $blue + "type the starting port : " + $blue
start_port = gets.chomp.to_i
print $blue + "type the ending port : " + $blue
end_port = gets.chomp.to_i

puts $orange + "scanning #{target_ip} from port #{start_port} to #{end_port}..." + $orange
scan_ports(target_ip, start_port, end_port) # scan to example 1 to 9000
